Sources/PackageCollectionsSigningLibc/ocsp_vfy.c

```diff
@@ -0,0 +1,353 @@
+    X509_VERIFY_PARAM *vp;
+    int ret = -1;
+
+    if (ctx == NULL) {
+        goto end;
+    }
+    if (!X509_STORE_CTX_init(ctx, st, signer, untrusted)) {
+        goto end;
+    }
+    if ((vp = X509_STORE_CTX_get0_param(ctx)) == NULL)
+        goto end;
+    if ((flags & OCSP_PARTIAL_CHAIN) != 0)
+        X509_VERIFY_PARAM_set_flags(vp, X509_V_FLAG_PARTIAL_CHAIN);
+    if (response
+            && X509_get_ext_by_NID(signer, NID_id_pkix_OCSP_noCheck, -1) >= 0)
+        /*
+         * Locally disable revocation status checking for OCSP responder cert.
+         * Done here for CRLs; TODO should be done also for OCSP-based checks.
+         */
+        X509_VERIFY_PARAM_clear_flags(vp, X509_V_FLAG_CRL_CHECK);
+    X509_STORE_CTX_set_purpose(ctx, X509_PURPOSE_OCSP_HELPER);
+    X509_STORE_CTX_set_trust(ctx, X509_TRUST_OCSP_REQUEST);
+    /* TODO: why is X509_TRUST_OCSP_REQUEST set? Seems to get ignored. */
+
+    ret = X509_verify_cert(ctx);
+    if (ret <= 0) {
+        goto end;
+    }
+    if (chain != NULL)
+        *chain = X509_STORE_CTX_get1_chain(ctx);
+
+ end:
+    X509_STORE_CTX_free(ctx);
+    return ret;
+}
+
+static int ocsp_verify(OCSP_BASICRESP *bs,
+                       X509 *signer, unsigned long flags)
+{
+    EVP_PKEY *skey;
+    int ret = 1;
+
+    if ((flags & OCSP_NOSIGS) == 0) {
+        // Note (Yim Lee): BoringSSL has X509_get_pubkey but not X509_get0_pubkey.
+        // The only difference between the two is that the key returned by X509_get_pubkey
+        // must be freed up after use. We choose to call X509_get_pubkey instead to avoid
+        // copying a bunch of code and risk complications.
+        if ((skey = X509_get_pubkey(signer)) == NULL) {
+            ret = -1;
+            goto end;
+        }
+        ret = OCSP_BASICRESP_verify(bs, skey);
+    }
+end:
+    EVP_PKEY_free(skey);
+    return ret;
+}
+
+/* Verify a basic response message */
+int OCSP_basic_verify(OCSP_BASICRESP *bs, STACK_OF(X509) *certs,
+                      X509_STORE *st, unsigned long flags)
+{
+    X509 *signer, *x;
+    STACK_OF(X509) *chain = NULL;
+    STACK_OF(X509) *untrusted = NULL;
+    int ret = ocsp_find_signer(&signer, bs, certs, flags);
+    int i;
+
+    if (ret == 0) {
+        goto end;
+    }
+    if ((ret == 2) && (flags & OCSP_TRUSTOTHER) != 0)
+        flags |= OCSP_NOVERIFY;
+
+    if ((ret = ocsp_verify(bs, signer, flags)) <= 0)
+        goto end;
+    if ((flags & OCSP_NOVERIFY) == 0) {
+        ret = -1;
+        if ((flags & OCSP_NOCHAIN) != 0) {
+            untrusted = NULL;
+        } else if (bs->certs != NULL && certs != NULL) {
+            untrusted = sk_X509_dup(bs->certs);
+            // Note (Yim Lee): BoringSSL doesn't have X509_add_certs. Since here we are just adding contents of
+            // one stack to another, we'd rather not copy a bunch of code from OpenSSL and risk complications.
+//            if (!X509_add_certs(untrusted, certs, X509_ADD_FLAG_DEFAULT))
```

Comment:
Similarly for this. Copying the functions _properly_ would involve `x509v3_cache_extensions` and I didn't feel comfortable with the idea at all. The usage here doesn't need any of the fancy flags so I change it to just call `sk_X509_insert` directly.